### DISCUSSION TITLE
shouldNotThrow* assertions should return value and not Unit

### DISCUSSION THREAD

**Discussion by Amejonah1200** [kotest/kotest#3972] [6.0 changes]
Hello o7
I'm new to kotest and found something bothering me: why do `shouldNotThrow*` (without Unit) not return a value?
It would be helpful when dealing with methods which instead of giving a Result<T>, they throw on error and return T.

Technically, I would say "same goes for tests too"...
```
val id = "gen id" { genId().shouldNotBeNull() }
```

**Comment by AlexCue987** [comment]
TBH I fail to see any practical need for this change. If we want to do anything with the returned value, we can just use it in an assertion, such as `myFunction() shouldBe 42`. And the reason  `shouldNotThrow` exists is to say: "this is not an unfinished test, if it doesn't blow up, it works correctly"

**Reply by Amejonah1200** [reply]
Main idea is to wrap exception-throwing functions and working with the value later.

```kt
"my test" {
  val id = shouldNotThrowAny { genId().shouldNotBeNull() }
  // vs
  var id: Int? = null
  shouldNotThrowAny {
    id = genId().shouldNotBeNull()
  }
}
```

**Reply by AlexCue987** [reply]
seems redundant to me, can be simplified to just:
```
"my test" {
  val id = genId().shouldNotBeNull() 
}
```

**Reply by Amejonah1200** [reply]
My main point was that the lambda can return any value, akin to https://junit.org/junit5/docs/5.3.0/api/org/junit/jupiter/api/Assertions.html#assertDoesNotThrow(org.junit.jupiter.api.function.ThrowingSupplier)
As the function ensures that there is no exception to be thrown (for example SQLException).

**Reply by AlexCue987** [reply]
Can you show us an example: a test without this fetature, and how that test would be improved by the feature you are proposing?
